compiler/src/lnntoamm/Fn.ts

```diff
@@ -0,0 +1,235 @@
+import Output from './Amm';
+import opcodes from './opcodes';
+import Scope from './Scope';
+import Stmt, { Assign, Dec, Emit, Exit, FnArg, MetaData, Ref } from './Statement';
+import Type, { FunctionType } from './Types';
+import { TODO } from './util';
+
+// the value is null if the type is to be inferred
+export type Args = {[name: string]: FnArg};
+
+export default class Fn {
+  // null if it's an anonymous fn
+  name: string | null
+  ast: LPNode
+  // the scope this function is defined in is the `par`
+  scope: Scope
+  args: Args
+  retTy: Type
+  body: Stmt[]
+  // not used by this class, but used by Statements
+  metadata: MetaData
+  fnType: FunctionType
+
+  get argNames(): string[] {
+    return Object.keys(this.args);
+  }
+
+  constructor(
+    ast: LPNode,
+    scope: Scope,
+    name: string | null,
+    args: Args,
+    retTy: Type | null,
+    body: Stmt[],
+    metadata: MetaData = null,
+  ) {
+    this.ast = ast;
+    this.scope = scope;
+    this.name = name;
+    this.args = args;
+    for (let argName of Object.keys(this.args)) {
+      if (this.args[argName].ty === null) {
+        this.args[argName].ty = Type.generate();
+      }
+    }
+    this.retTy = retTy !== null ? retTy : Type.generate();
+    this.body = body;
+    this.metadata = metadata !== null ? metadata : new MetaData(scope);
+    this.fnType = new FunctionType(
+      this.name,
+      Object.values(this.args).map(a => a.ty),
+      this.retTy
+    );
+  }
+
+  static fromFunctionsAst(
+    ast: LPNode,
+    scope: Scope,
+    // metadata: MetaData = null,
+  ): Fn {
+    // TODO: inheritance
+    let metadata = new MetaData(scope);
+
+    let work = ast;
+    const name = work.get('optname').has() ? work.get('optname').get().t : null;
+    let args: Args = {};
+    if (work.get('optargs').has('arglist')) {
+      let argAsts = [
+        work.get('optargs').get('arglist'),
+        ...work.get('optargs').get('arglist').get('cdr').getAll(),
+      ];
+      argAsts.forEach(argAst => {
+        const arg = FnArg.fromArgAst(argAst, metadata)
+        args[arg.name] = arg;
+      });
+    }
+
+    const retTy = work.get('optreturntype').has() ? work.get('optreturntype').get().get('fulltypename') : 'void';
+
+    let body = [];
+    let bodyAsts: LPNode | LPNode[] = work.get('fullfunctionbody');
+    if (bodyAsts.has('functionbody')) {
+      bodyAsts = bodyAsts.get('functionbody').get('statements').getAll().map(s => s.get('statement'));
+      bodyAsts.forEach(ast => body.push(...Stmt.fromAst(ast, metadata)));
+    } else {
+      bodyAsts = bodyAsts.get('assignfunction');
+      body = Stmt.fromAst(bodyAsts, metadata);
+      const retVal = body[body.length - 1];
+      if (!(retVal instanceof Dec)) {
+        throw new Error(`illegal function body: ${bodyAsts}`);
+      }
+      body.push(new Exit(bodyAsts, retVal.ref()));
+    }
+
+    return new Fn(
+      ast,
+      new Scope(scope),
+      name,
+      args,
+      Type.getFromTypename(retTy, scope),
```

Comment:
This breaks return type inference, doesn't it? (I didn't see the return type being taken from the exit statement(s) so it would be `void` if not explicitly defined.

Reply:
the exit statements do contribute to the return type (look at `Exit.fromAst` - you'll see `metadata.retConstraints.push(...)`). However, you're right - instead of defaulting to `void` when there's no return type specified, I should default to `Type.oneOf([Type.generate(), opcodes().get('void')])` so that it prefers the `void` type (since there is no return) but will get constrained to a non-void return type once the constraints get processed.